src/main/java/net/dv8tion/jda/internal/entities/GuildImpl.java

```diff
@@ -781,6 +785,57 @@ public boolean isAvailable()
                 getJDA().getEntityBuilder().createMember(this, resp.getObject()));
     }
 
+    @Nonnull
+    @Override
+    public CompletableFuture<MemberPresence> retrieveMemberPresence(long memberId)
+    {
+        Member member = getMemberById(memberId);
+        if (member != null
+            && api.isCacheFlagSet(CacheFlag.ACTIVITY)
+            && api.isCacheFlagSet(CacheFlag.CLIENT_STATUS))
+            return CompletableFuture.completedFuture(new MemberPresenceImpl(member));
```

Comment:
Why do we need to recreate a MemberPresence here when Member itself implements the interface?

Reply:
I think its better to pass a snapshot to the future as the member state might change before the future is runing callbacks. The `toString` is also the same for both cases this way.